include/yaclib/coroutine/detail/future_awaiter.hpp

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <yaclib/async/future.hpp>
+#include <yaclib/coroutine/detail/coroutine.hpp>
+#include <yaclib/util/detail/atomic_counter.hpp>
+#include <yaclib/util/type_traits.hpp>
+
+namespace yaclib::detail {
+
+class FutureAwaiter final {
+ public:
+  FutureAwaiter() = delete;
+  FutureAwaiter(const FutureAwaiter&) = delete;
+  FutureAwaiter(FutureAwaiter&&) = delete;
+  FutureAwaiter& operator=(FutureAwaiter&&) = delete;
+  FutureAwaiter& operator=(const FutureAwaiter&) = delete;
+  ~FutureAwaiter() = default;
+
+  template <typename... V, typename... E>
+  explicit FutureAwaiter(Future<V, E>&... fs) : _await_core{1 + sizeof...(fs)} {
+    auto wait_count = (... + static_cast<std::size_t>(fs.GetCore()->SetWait(_await_core)));
+    _await_core.SubEqual(sizeof...(fs) - wait_count);
+  }
+
+  template <typename It, typename RangeIt>
+  explicit FutureAwaiter(It it, RangeIt begin, RangeIt end) : _await_core{static_cast<std::size_t>(1 + end - begin)} {
```

Comment:
This constructor ahoud be rewritten as Iterator and size.
I think better when user use std::distance, not library, see comment in other same places in yaclib.
Because of often user doesn't have random access iterators but have size of range

Reply:
Also this is detail, and user should not use this code